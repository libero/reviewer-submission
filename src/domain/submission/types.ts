import { uuidType } from 'typesafe-uuid';
import File from '../file/services/models/file';

export class SubmissionId extends uuidType<'SubmissionId'>() {}

export type AuthorDetails = {
    firstName: string;
    lastName: string;
    email: string;
    institution: string;
};

export type ManuscriptDetails = {
    title?: string;
    subjects?: string[];
    previouslyDiscussed?: string;
    previouslySubmitted?: string;
    cosubmission?: string[];
};

export type DisclosureDetails = {
    submitterSignature?: string;
    disclosureConsent?: boolean;
};

export type ReviewerAlias = {
    name: string;
    email: string;
};

export type OpposedReviewer = {
    name: string;
    email: string;
};

export type EditorDetails = {
    suggestedSeniorEditors?: Array<string>;
    opposedSeniorEditors?: Array<string>;
    opposedSeniorEditorsReason?: string;
    suggestedReviewingEditors?: Array<string>;
    opposedReviewingEditors?: Array<string>;
    opposedReviewingEditorsReason?: string;
    suggestedReviewers?: Array<ReviewerAlias>;
    opposedReviewers?: Array<OpposedReviewer>;
    opposedReviewersReason?: string;
};

export type FileDetails = {
    coverLetter?: string;
    manuscriptFile?: File | null; // responsibility of the Files Service
    supportingFiles?: Array<File>; // responsibility of the Files Service
};

export enum SubmissionStatus {
    INITIAL = 'INITIAL',
    MECA_EXPORT_PENDING = 'MECA_EXPORT_PENDING',
    MECA_EXPORT_FAILED = 'MECA_EXPORT_FAILED',
    MECA_EXPORT_SUCCEEDED = 'MECA_EXPORT_SUCCEEDED',
    MECA_IMPORT_FAILED = 'MECA_IMPORT_FAILED',
    MECA_IMPORT_SUCCEEDED = 'MECA_IMPORT_SUCCEEDED',
}
