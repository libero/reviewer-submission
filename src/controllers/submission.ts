import { Logger, HttpException, HttpStatus } from '@nestjs/common';
import { Option, None } from 'funfix';
import { SubmissionRepository, SubmissionId, Submission } from '../types/submission';

// TODO: this is temporary until the picture on validation and article types is clearer.
const articlesTypes = ['researchArticle', 'featureArticle', 'researchAdvance'];

export class SubmissionController {
    private readonly logger = new Logger(SubmissionController.name);

    constructor(readonly repository: SubmissionRepository) {}

    close(): void {
        this.logger.log('Closing repository');
        this.repository.close();
    }

    async findAll(): Promise<Option<Submission[]>> {
        return new Promise(async resolve => {
            const result: Option<Submission[]> = await this.repository.findAll();
            if (result.isEmpty()) {
                resolve(None);
            } else {
                resolve(Option.of(result.get()));
            }
        });
    }

    async create(articleType: string): Promise<Option<Submission>> {
        if (!articlesTypes.includes(articleType)) {
            throw new HttpException('Unprocessable entity', HttpStatus.UNPROCESSABLE_ENTITY);
        }
        return this.repository.create(articleType);
    }

    async findOne(id: SubmissionId): Promise<Option<Submission>> {
        return this.repository.findById(id);
    }

    async changeTitle(id: SubmissionId, title: string): Promise<Option<Submission>> {
        return this.repository.changeTitle(id, title);
    }

    async delete(id: SubmissionId): Promise<number> {
        return this.repository.delete(id);
    }
}
