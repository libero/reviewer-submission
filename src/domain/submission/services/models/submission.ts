import { SubmissionId } from '../../types';
import File from '../../../file/services/models/file';
import { FileId } from 'src/domain/file/types';

export enum ArticleType {
    RESEARCH_ARTICLE = 'researchArticle',
    FEATURE_ARTICLE = 'featureArticle',
    RESEARCH_ADVANCE = 'researchAdvance',
}

export enum SubmissionStatus {
    INITIAL = 'INITIAL',
    MECA_EXPORT_PENDING = 'MECA_EXPORT_PENDING',
    MECA_EXPORT_FAILED = 'MECA_EXPORT_FAILED',
    MECA_EXPORT_SUCCEEDED = 'MECA_EXPORT_SUCCEEDED',
    MECA_IMPORT_FAILED = 'MECA_IMPORT_FAILED',
    MECA_IMPORT_SUCCEEDED = 'MECA_IMPORT_SUCCEEDED',
}

export default class Submission {
    id: SubmissionId;
    title: string;
    updated: Date;
    articleType: ArticleType;
    status: string;
    createdBy: string;
    manuscriptFile?: File | null;
    supportingFiles?: Array<File>;

    // This is wired up so that you can create an entity from the DTO described by ISubmission
    constructor({
        id,
        title,
        updated,
        articleType,
        status,
        createdBy,
        manuscriptFile,
        supportingFiles,
    }: {
        id: SubmissionId;
        title: string;
        updated?: Date;
        articleType: string;
        status: string;
        createdBy: string;
        manuscriptFile?: File | null;
        supportingFiles?: Array<File>;
    }) {
        this.id = id;
        this.title = title;
        this.updated = updated || new Date();
        this.articleType = this.articleTypeFromString(articleType);
        this.status = status;
        this.createdBy = createdBy;
        this.manuscriptFile = manuscriptFile;
        this.supportingFiles = supportingFiles;
    }

    private articleTypeFromString(type: string): ArticleType {
        switch (type) {
            case 'researchArticle':
                return ArticleType.RESEARCH_ARTICLE;
            case 'featureArticle':
                return ArticleType.FEATURE_ARTICLE;
            case 'researchAdvance':
                return ArticleType.RESEARCH_ADVANCE;
            default:
                throw new Error('Invalid article type');
        }
    }

    public setManuscriptFile(fileId: FileId, filename: string, mimeType: string, fileSize: number): void {
        if (!this.manuscriptFile) {
            throw new Error('Manuscript file already present');
        }

        this.manuscriptFile = File.makeManuscriptFile(fileId, this.id, filename, mimeType, fileSize);
    }

    public getManuscriptFile(): File | null {
        if (!this.manuscriptFile) {
            return null;
        }

        return this.manuscriptFile;
    }

    public setManuscriptFileStatusToStored(): void {
        if (this.manuscriptFile) {
            this.manuscriptFile.setStatusToStored();
        }
    }

    public setManuscriptFileStatusToCancelled(): void {
        if (this.manuscriptFile) {
            this.manuscriptFile.setStatusToCancelled();
        }
    }
}
