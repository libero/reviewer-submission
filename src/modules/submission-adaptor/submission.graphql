type Submission {
  id: ID!
  title: String!
  updated: String!
}

type Query {
  getSubmissions: [Submission!]!
  getSubmission(id: ID!): Submission
}

type Mutation {
  startSubmission(articleType: String!): Submission!
  changeSubmissionTitle(id: ID!, title: String!): Submission!
  deleteSubmission(id: ID!): Boolean
}
