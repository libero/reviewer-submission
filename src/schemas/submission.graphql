scalar Upload

type Submission {
    id: ID!
    title: String!
    updated: String!
    articleType: String!
    author: AuthorDetails!
    manuscriptFile: File
    supportingFiles: [File]
    coverLetter: String
    suggestions: [String]!
}

type File {
    id: ID!
    created: DateTime!
    updated: DateTime
    type: String
    label: String
    filename: String
    url: String
    mimeType: String
    size: Int
    lastChangeDate: DateTime!
    lastChangeUser: User!
    status: String!
    downloadLink: String!
}

type AuthorDetails {
    firstName: String!
    lastName: String!
    email: String!
    institution: String!
}

input AuthorDetailsInput {
    firstName: String!
    lastName: String!
    email: String!
    institution: String!
}
