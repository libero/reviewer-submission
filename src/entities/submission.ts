import { SubmissionId, DtoSubmission, DtoViewSubmission, Submission } from '../types/submission';

export class SubmissionEntity implements Submission {
    id: SubmissionId;

    title: string;

    updated: Date;

    articleType: string;

    // This is wired up so that you can create an entity from the DTO described by ISubmission
    constructor({
        id,
        title,
        updated,
        articleType,
    }: {
        id: SubmissionId;
        title: string;
        updated?: Date;
        articleType: string;
    }) {
        this.id = id;
        this.title = title;
        this.updated = updated || new Date();
        this.articleType = articleType;
    }
}

export class SubmissionMapper {
    public static toDto(sub: Submission): DtoSubmission {
        return {
            id: sub.id,
            title: sub.title,
            updated: sub.updated,
            articleType: sub.articleType,
        };
    }
    public static toViewDto(sub: Submission): DtoViewSubmission {
        return {
            id: sub.id,
            title: sub.title,
            updated: sub.updated,
            articleType: sub.articleType,
        };
    }
    public static fromDto(sub: DtoSubmission): SubmissionEntity {
        return new SubmissionEntity(sub);
    }
}
