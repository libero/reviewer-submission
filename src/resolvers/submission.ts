import { SubmissionService } from '../services/submission';
import { SubmissionId, DtoViewSubmission } from '../types/submission';
import { IResolvers } from 'apollo-server-express';

const resolvers = (submissionService: SubmissionService): IResolvers => ({
    Query: {
        async getSubmissions(): Promise<DtoViewSubmission[] | null> {
            return await submissionService.findAll();
        },
        async getSubmission(id: string): Promise<DtoViewSubmission | null> {
            return await submissionService.findOne(SubmissionId.fromUuid(id));
        },
    },
    Mutation: {
        async startSubmission(_, args: { articleType: string }): Promise<DtoViewSubmission | null> {
            return await submissionService.create(args.articleType);
        },

        async changeSubmissionTitle(_, args: { id: string; title: string }): Promise<DtoViewSubmission | null> {
            return await submissionService.changeTitle(SubmissionId.fromUuid(args.id), args.title);
        },

        async deleteSubmission(_, args: { id: SubmissionId }): Promise<boolean> {
            return await submissionService.delete(args.id);
        },
    },
});

export const SubmissionResolvers = resolvers;
