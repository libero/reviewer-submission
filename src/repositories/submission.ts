import { SubmissionRepository, DtoSubmission, SubmissionId, Submission } from '../types/submission';
import * as Knex from 'knex';
import { Logger } from '@nestjs/common';
import { SubmissionMapper } from '../entities/submission';

export class KnexSubmissionRepository implements SubmissionRepository {
    private readonly TABLE_NAME = 'submission';
    private readonly logger = new Logger(KnexSubmissionRepository.name);

    public constructor(private readonly knex: Knex<{}, unknown[]>) {}

    public async initSchema(): Promise<boolean | void> {
        // TODO: Add a method for handling when the table does/doesn't exist
        // as this will error if the table already exists
        // XXX: Maybe move this to migrations
        const hasTable = await this.knex.schema.hasTable(this.TABLE_NAME);

        if (hasTable) {
            return hasTable;
        }

        return await this.knex.schema.createTable(this.TABLE_NAME, (table: Knex.CreateTableBuilder) => {
            table.uuid('id');
            table.string('title');
            table.timestamp('updated').defaultTo(this.knex.fn.now());
            this.logger.log(`created table ${this.TABLE_NAME}`);
        });
    }

    close(): void {
        this.logger.log(`Closing repository.`);
        this.knex.destroy();
    }

    public async findAll(): Promise<Submission[]> {
        const result = await this.knex(this.TABLE_NAME).select<DtoSubmission[]>('id', 'title', 'updated');
        return result.map(SubmissionMapper.fromDto);
    }

    public async findById(id: SubmissionId): Promise<Submission | null> {
        const rows = await this.knex(this.TABLE_NAME)
            .where({ id })
            .select<DtoSubmission[]>('id', 'title', 'updated');

        return rows.length > 0 ? SubmissionMapper.fromDto(rows[0]) : null;
    }

    public async save(sub: Submission): Promise<Submission | null> {
        const dtoSubmission: DtoSubmission = SubmissionMapper.toDto(sub);
        dtoSubmission.updated = new Date();
        await this.knex(this.TABLE_NAME)
            .insert(dtoSubmission)
            .returning('id');
        return SubmissionMapper.fromDto(dtoSubmission);
    }

    public async delete(id: SubmissionId): Promise<boolean> {
        const res = await this.knex(this.TABLE_NAME)
            .where({ id })
            .delete();
        return res >= 1 ? true : false;
    }
}
