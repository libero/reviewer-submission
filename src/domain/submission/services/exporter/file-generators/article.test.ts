import { mocked } from 'ts-jest/utils';
import Axios from 'axios';
import * as crypto from 'crypto';
import { replace, includes } from 'lodash';
import ArticleGenerator from './article';
import submission from './article.test.data';

jest.mock('axios');

const md5 = (s: string): string =>
    crypto
        .createHash('md5')
        .update(s)
        .digest('hex');

const obfuscate = (s: string, re: RegExp): string => {
    let m;
    const matches = [];
    do {
        m = re.exec(s);
        if (m) {
            matches.push({ str: m[1], index: m.index });
        }
    } while (m);

    let newStr = s;

    matches.forEach(item => {
        // make sure this is idempotent
        if (includes(item.str, '.') && includes(item.str, '@')) {
            newStr = replace(newStr, item.str, md5(item.str));
        }
    });
    return newStr;
};

const obfuscateEmail = (s: string): string => {
    const re = new RegExp('<email>([^<>]+)</email>', 'g');
    return obfuscate(s, re);
};

const editorsData: { [key: string]: {} } = {
    '1e9e661f': {
        name: {
            surname: 'Zoghbi',
            givenNames: 'Huda',
        },
        emailAddresses: [{ value: '8ac3df4aa667c9217fe4d8bedfc06482', access: 'restricted' }],
        affiliations: [
            {
                name: ['Texas Children\u0027s Hospital'],
            },
        ],
    },
    '3edb2ed8': {
        name: {
            surname: 'Li',
            givenNames: 'Wenhui',
        },
        emailAddresses: [{ value: '5ddd85b12237e714f490a9c6cf585566', access: 'restricted' }],
        affiliations: [
            {
                name: ['National Institute of Biological Sciences'],
            },
        ],
    },
    '232d9893': {
        name: {
            surname: 'Struhl',
            givenNames: 'Kevin',
            preferred: 'Kevin Struhl',
            index: 'Struhl, Kevin',
        },
        emailAddresses: [{ value: '4f5cd09b1742dad17ffe48d0c33def8f', access: 'restricted' }],
        affiliations: [{ name: ['Harvard Medical School'] }],
    },
    '87f34696': {
        name: { surname: 'Yu', givenNames: 'Hao', preferred: 'Hao Yu', index: 'Yu, Hao' },
        emailAddresses: [{ value: 'd3fa5c7081fecbc7cfc02be293834b86', access: 'restricted' }],
        affiliations: [{ name: ['National University of Singapore \u0026 Temasek Life Sciences Laboratory'] }],
    },
    fd8295ba: {
        name: { surname: 'Zhang', givenNames: 'Hong', preferred: 'Hong Zhang', index: 'Zhang, Hong' },
        emailAddresses: [{ value: 'e69bf9f08f18063b3098e1cd63c26a23', access: 'restricted' }],
        affiliations: [
            {
                name: ['Institute of Biophysics Chinese Academy of Sciences'],
                address: { formatted: ['China'], components: { country: 'China' } },
            },
        ],
    },
    '6fabd619': {
        name: { surname: 'Zilberman', givenNames: 'Daniel', preferred: 'Daniel Zilberman', index: 'Zilberman, Daniel' },
        emailAddresses: [{ value: '08c6488ad00599649606fbb926a63f12', access: 'restricted' }],
        affiliations: [
            {
                name: ['John Innes Centre'],
                address: { formatted: ['United Kingdom'], components: { country: 'United Kingdom' } },
            },
        ],
    },
};

describe('ArticleGenerator', () => {
    it('generates correct xml', async () => {
        const mockedGet = mocked(Axios.get, true);
        mockedGet.mockImplementation(query => {
            const id = query.split('/').pop();
            return Promise.resolve({
                data: id ? editorsData[id] : {},
            });
        });

        const articleGenerator = new ArticleGenerator(submission);
        const output = await articleGenerator.execute();

        expect(obfuscateEmail(output)).toMatchSnapshot();
    });
});
