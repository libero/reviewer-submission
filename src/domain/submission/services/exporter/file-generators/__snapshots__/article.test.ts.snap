// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleGenerator generates correct xml 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<!DOCTYPE article SYSTEM \\"JATS-journalpublishing1.dtd\\">
<article>
  <front>
    <journal-meta>
      <journal-id journal-id-type=\\"pmc\\"/>
      <journal-id journal-id-type=\\"publisher\\"/>
      <issn/>
      <publisher>
        <publisher-name/>
      </publisher>
    </journal-meta>
    <article-meta>
      <article-id pub-id-type=\\"manuscript\\">604e06ca-882d-4b5b-a147-e016893e60e9</article-id>
      <article-categories>
        <subj-group subj-group-type=\\"article_type\\">
          <subject>5</subject>
        </subj-group>
        <subj-group>
          <subj-group subj-group-type=\\"subject_areas\\">
            <subject>Biochemistry and Chemical Biology</subject>
          </subj-group>
        </subj-group>
        <subj-group>
          <subj-group subj-group-type=\\"subject_areas\\">
            <subject>Chromosomes and Gene Expression</subject>
          </subj-group>
        </subj-group>
      </article-categories>
      <title-group>
        <article-title>Test Manuscript</article-title>
      </title-group>
      <contrib-group content-type=\\"authors\\">
        <contrib contrib-type=\\"author\\" corresp=\\"yes\\">
          <name>
            <surname>User</surname>
            <given-names>Test</given-names>
          </name>
          <email>test.user@example.com</email>
          <xref ref-type=\\"aff\\" rid=\\"aff6\\"/>
        </contrib>
      </contrib-group>
      <contrib-group content-type=\\"potential_senior_editors\\">
        <contrib contrib-type=\\"suggested_senior_editor\\">
          <name>
            <surname>Bluth</surname>
            <given-names>Michael</given-names>
          </name>
          <email>m.bluth@example.com</email>
          <xref ref-type=\\"aff\\" rid=\\"aff1\\"/>
        </contrib>
        <contrib contrib-type=\\"suggested_senior_editor\\">
          <name>
            <surname>Holt</surname>
            <given-names>Steve</given-names>
          </name>
          <email>s.holt@example.com</email>
          <xref ref-type=\\"aff\\" rid=\\"aff2\\"/>
        </contrib>
        <contrib contrib-type=\\"opposed_senior_editor\\">
          <name>
            <surname>Funke</surname>
            <given-names>Tobias</given-names>
          </name>
          <email>t.funke@example.com</email>
          <xref ref-type=\\"aff\\" rid=\\"aff3\\"/>
        </contrib>
      </contrib-group>
      <contrib-group content-type=\\"potential_reviewing_editors\\">
        <contrib contrib-type=\\"suggested_reviewing_editor\\">
          <name>
            <surname>Funke</surname>
            <given-names>Lindsay</given-names>
          </name>
          <email>l.funke@example.com</email>
          <xref ref-type=\\"aff\\" rid=\\"aff4\\"/>
        </contrib>
        <contrib contrib-type=\\"suggested_reviewing_editor\\">
          <name>
            <surname>Bluth</surname>
            <given-names>Lucille</given-names>
          </name>
          <email>l.bluth@example.com</email>
          <xref ref-type=\\"aff\\" rid=\\"aff5\\"/>
        </contrib>
        <contrib contrib-type=\\"opposed_reviewing_editor\\">
          <name>
            <surname>Bluth</surname>
            <given-names>Gob</given-names>
          </name>
          <email>g.bluth@example.com</email>
          <xref ref-type=\\"aff\\" rid=\\"aff0\\"/>
        </contrib>
      </contrib-group>
      <contrib-group content-type=\\"potential_reviewers\\">
        <contrib contrib-type=\\"suggested_reviewer\\">
          <name>
            <surname>Reviewer</surname>
            <given-names>J. Edward</given-names>
          </name>
          <email>edward@example.com</email>
        </contrib>
        <contrib contrib-type=\\"suggested_reviewer\\">
          <name>
            <surname>de Reviewer</surname>
            <given-names>Frances</given-names>
          </name>
          <email>frances@example.org</email>
        </contrib>
        <contrib contrib-type=\\"suggested_reviewer\\">
          <name>
            <surname>Reviewer</surname>
            <given-names>Gertrude</given-names>
          </name>
          <email>gertrude@example.net</email>
        </contrib>
        <contrib contrib-type=\\"opposed_reviewer\\">
          <name>
            <surname>Hoo</surname>
            <given-names/>
          </name>
          <email>laughing@example.com</email>
        </contrib>
        <contrib contrib-type=\\"opposed_reviewer\\">
          <name>
            <surname>Who</surname>
            <given-names/>
          </name>
          <email>singing@example.com</email>
        </contrib>
      </contrib-group>
      <aff id=\\"aff0\\">
        <institution>Magician circle</institution>
      </aff>
      <aff id=\\"aff1\\">
        <institution>Bluth research centre</institution>
      </aff>
      <aff id=\\"aff2\\">
        <institution>High School</institution>
      </aff>
      <aff id=\\"aff3\\">
        <institution>Blue man group</institution>
      </aff>
      <aff id=\\"aff4\\">
        <institution>Home</institution>
      </aff>
      <aff id=\\"aff5\\">
        <institution>Balboa Towers</institution>
      </aff>
      <aff id=\\"aff6\\">
        <institution>University of eLife</institution>
      </aff>
      <author-notes>
        <fn>
          <p/>
        </fn>
      </author-notes>
      <pub-date>
        <day/>
        <month/>
        <year/>
      </pub-date>
      <related-article related-article-type=\\"companion\\">
        <article-title>asdassss</article-title>
      </related-article>
      <related-article related-article-type=\\"companion\\">
        <article-title>Another</article-title>
      </related-article>
      <custom-meta-group>
        <custom-meta>
          <meta-name>opposed-senior-editor-comment</meta-name>
          <meta-value>From another galaxy</meta-value>
        </custom-meta>
        <custom-meta>
          <meta-name>opposed-reviewing-editor-comment</meta-name>
          <meta-value>Not any more</meta-value>
        </custom-meta>
        <custom-meta>
          <meta-name>opposed-reviewer-comment</meta-name>
          <meta-value>Wandering days are over</meta-value>
        </custom-meta>
        <custom-meta>
          <meta-name>previously-discussed</meta-name>
          <meta-value>Talked to bob about it</meta-value>
        </custom-meta>
        <custom-meta>
          <meta-name>previously-submitted</meta-name>
          <meta-value>Original Test Title</meta-value>
        </custom-meta>
        <custom-meta>
          <meta-name>created-by</meta-name>
          <meta-value>6d8cd1ce-15b6-46c1-b901-bc91598c8f2d</meta-value>
        </custom-meta>
      </custom-meta-group>
    </article-meta>
  </front>
</article>"
`;
